File: Problems/029-divide-two-integers/divide-two-integers.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        sign = (dividend > 0) ^ (divisor > 0)
        dividend, divisor = abs(dividend), abs(divisor)

        count = 0
        while dividend >= divisor:
            count += 1
            divisor <<= 1
        res = 0
        while count > 0:
            count -= 1
            divisor >>= 1
            if divisor <= dividend:
                res += 1 << count
                dividend -= divisor

        if sign: res = -res
        return res if -(1 << 31) <= res <= (1 << 31) - 1 else (1 << 31) - 1

File: Problems/029-divide-two-integers/divide-two-integers.go
package problem029

func divide(dividend int, divisor int) int {
	sign1, dividend := abs(dividend)
	sign2, divisor := abs(divisor)

	var cnt uint
	for divisor <= dividend {
		cnt++
		divisor <<= 1
	}

	res := 0
	for cnt > 0 {
		cnt--
		divisor >>= 1
		if divisor <= dividend {
			res += 1 << cnt
			dividend -= divisor
		}
	}

	if sign1 != sign2 {
		res = -res
	}

	if res < -(1<<31) || res > (1<<31)-1 {
		return 1<<31 - 1
	}
	return res
}

func abs(n int) (sign, num int) {
	sign = 1
	num = n
	if n < 0 {
		sign, num = -1, -n
	}
	return
}
